### PR TITLE
Users have different CRUD rights depending on their role

### PR DESCRIPTION
**Pull request by Fleurtam**
#### Description
Changes proposed in this pull request:

* Users should have different CRUD rights  according to their account type

#### What I have learned working on this feature:
Learning to work with pundit 

#### Screenshots:
screen shot 2018-03-21 at 15 42 34

### CODE REVIEW

File: app/controllers/articles_controller.rb [hunk 1]
```diff
@@ -1,9 +1,23 @@
...
+    authorize @article
+  end
+
+  def show
+    @article = Article.find_by(id: params[:id])
+    @categories = Category.all
+    authorize @article
+  end
+
+
```
Comment (davidboll) [inline app/controllers/articles_controller.rb:14]: Remove empty line

File: app/models/user.rb [hunk 2]
```diff
@@ -1,4 +1,22 @@
...
+
+    def admin?
+      self.role == 'admin'
+    end
+
+    def subscriber?
+      self.role == 'subscriber'
+    end
+
+
```
Comment (davidboll) [inline app/models/user.rb]: Remove empty line

File: app/policies/article_policy.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  def create?
+    new?
+  end
+
+  def destroy?
+    new?
+  end
+
+  def show?
+    @user.author? || @user.subscriber?
```
Comment (davidboll) [inline app/policies/article_policy.rb]: Maybe add @user.admin?

File: features/different_users_have_different_capabilities.feature [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  In order to control which type of user is authorized to do what
+  I want to be able to set authorization status to different account types
+
+  Background:
+    Given following user exist
+      | email                 | role       |
+      | visitor@email.com     | visitor    |
+      | subscriber@email.com  | subscriber |
+      | author@email.com      | author     |
+      | admin@email.com       | admin      |
```
Comment (davidboll) [inline features/different_users_have_different_capabilities.feature:12]: If you are not testing the admin, remove it from background

File: features/user_can_delete_article.feature [hunk 5]
```diff
@@ -4,12 +4,20 @@ Feature: User should be able to delete articles
...
-    Given the following articles exist
-    | title               | body                  |
-    | Holger is the best  | Dreamteam for the win |
+    Given following user exist
+      | email                 | role       |
+      | visitor@email.com     | visitor    |
+      | subscriber@email.com  | subscriber |
+      | author@email.com      | author     |
+      | admin@email.com       | admin      |
+
```
Comment (davidboll) [inline features/user_can_delete_article.feature:13]: You are only testing author, remove other background-users.

File: features/user_can_make_comments.feature [hunk 6]
```diff
@@ -4,17 +4,26 @@ Feature: User should be able to add a comment
...
-    Given the following articles exist
-    | title               | body                  |
-    | Holger is the best  | Dreamteam for the win |
+    Given following user exist
+      | email                 | role       |
+      | visitor@email.com     | visitor    |
+      | subscriber@email.com  | subscriber |
+      | author@email.com      | author     |
+      | admin@email.com       | admin      |
+
```
Comment (davidboll) [inline features/user_can_make_comments.feature:13]: You are only testing author, remove other background-users.

File: features/user_can_see_articles.feature [hunk 7]
```diff
@@ -4,11 +4,25 @@ Feature: User should be able to see articles
...
-    Given the following articles exist
-    | title               | body                  |
-    | Holger is the best  | Dreamteam for the win |
+    Given following user exist
+      | email                 | role       |
+      | visitor@email.com     | visitor    |
+      | subscriber@email.com  | subscriber |
+      | author@email.com      | author     |
+      | admin@email.com       | admin      |
+
```
Comment (davidboll) [inline features/user_can_see_articles.feature:13]: You are only testing subscriber and visitor, remove other background-users.

File: features/user_should_be_able_to_create_articles.feature [hunk 8]
```diff
@@ -7,9 +7,16 @@ Feature: User should be able to create articles
     Given the following categories exist
       | name    |
       | Fashion |
+    And following user exist
+      | email                 | role       |
+      | visitor@email.com     | visitor    |
+      | subscriber@email.com  | subscriber |
+      | author@email.com      | author     |
+      | admin@email.com       | admin      |
+    And I am logged in as 'author@email.com'
```
Comment (davidboll) [inline features/user_should_be_able_to_create_articles.feature:16]: Same here.. remove other users from background

File: features/user_should_be_able_to_edit_articles.feature [hunk 9]
```diff
@@ -4,18 +4,27 @@ Feature: User should have the ability to edit articles
...
 
   Background:
-    Given the following articles exist
+    Given following user exist
+      | email                 | role       |
+      | visitor@email.com     | visitor    |
+      | subscriber@email.com  | subscriber |
+      | author@email.com      | author     |
+      | admin@email.com       | admin      |
+
```
Comment (davidboll) [inline features/user_should_be_able_to_edit_articles.feature:13]: Same here.

**Review by davidboll — CHANGES_REQUESTED** [review]
Looks good, but please clean up the code, se my comments :)

**Review by davidboll — DISMISSED** [review]

File: app/controllers/application_controller.rb [hunk 10]
```diff
@@ -1,2 +1,10 @@
 class ApplicationController < ActionController::Base
+  include Pundit
+
+  rescue_from Pundit::NotAuthorizedError, with: :user_not_authorized
+
+  def user_not_authorized
+    redirect_to root_path, notice: 'Aja baja!'
+  end
+
```
Comment (jyuw) [inline app/controllers/application_controller.rb]: remove this linebreak between the two ends

File: app/controllers/articles_controller.rb [hunk 11]
```diff
@@ -1,9 +1,22 @@
...
+    @article = Article.find_by(id: params[:id])
+    @categories = Category.all
+    authorize @article
+  end
+
   def create
     params[:article][:categories].shift
     @article = Article.new(article_params)
+    authorize @article
+
```
Comment (jyuw) [inline app/controllers/articles_controller.rb]: remove line break please

File: app/policies/article_policy.rb [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+  end
+
+  def destroy?
+    new?
+  end
+
+  def show?
+    @user.author? || @user.subscriber? || @user.admin?
+  end
+
```
Comment (jyuw) [inline app/policies/article_policy.rb]: this one too please

File: features/articles_belong_to_categories.feature [hunk 13]
```diff
@@ -4,9 +4,13 @@ Feature: Article can be linked to a category
   I would like to be able to categorize articles
 
   Background:
-    Given the following categories exist
+    Given following user exist
+      | email                 | role       |
+      | author@email.com      | author     |
```
Comment (jyuw) [inline features/articles_belong_to_categories.feature:9]: add line break between each given

File: features/articles_display_under_category_pages.feature [hunk 14]
```diff
@@ -4,17 +4,21 @@ Feature: Article are displayed under correct category
   I would like to be able to see the articles in separate category page views
 
   Background:
-    Given the following categories exist
+    Given following user exist
+      | email                 | role       |
+      | author@email.com      | author     |
```
Comment (jyuw) [inline features/articles_display_under_category_pages.feature:9]: add linebreak after each given

File: features/articles_display_under_category_pages.feature [hunk 15]
```diff
@@ -4,17 +4,21 @@ Feature: Article are displayed under correct category
...
       | name    |
       | Fashion |
+    And I am logged in as 'author@email.com'
 
   Scenario: User can create an article and see it display under category
     Given I am on the "Index" page
     When I click on "Write an Article"
     And I select "Fashion" from "Categories"
-    And I fill in "Title" with "Holger is the best"
+    And I fill in "Title" with "David is the best"
```
Comment (jyuw) [inline features/articles_display_under_category_pages.feature:21]: Please provide evidence of this claim

File: features/different_users_have_different_capabilities.feature [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+  Background:
+    Given following user exist
+      | email                 | role       |
+      | visitor@email.com     | visitor    |
+      | subscriber@email.com  | subscriber |
+      | author@email.com      | author     |
+    And the following categories exist
+      | name    |
+      | Fashion |
+    Given the following articles exist
```
Comment (jyuw) [inline features/different_users_have_different_capabilities.feature]: Linebreaks and change this to `And`

File: features/step_definitions/basic_steps.rb [hunk 17]
```diff
@@ -31,11 +36,22 @@
   select(category_name, from: category_list.downcase)
 end
 
-
 Then("I should be on the {string} page") do |article_title|
   expect(page.current_path).to eq find_article(article_title)
 end
 
+Given("I try to visit the {string} article-page") do |article_title|
```
Comment (jyuw) [inline features/step_definitions/basic_steps.rb]: Reuse step on line 25 using the pipes like `try to||am on` (I think)

File: features/user_can_delete_article.feature [hunk 18]
```diff
@@ -4,12 +4,16 @@ Feature: User should be able to delete articles
   I would like to be able to delete articles
 
   Background:
-    Given the following articles exist
-    | title               | body                  |
-    | Holger is the best  | Dreamteam for the win |
+    Given following user exist
```
Comment (jyuw) [inline features/user_can_delete_article.feature:7]: Line breaks here too please

File: features/user_can_make_comments.feature [hunk 19]
```diff
@@ -4,17 +4,21 @@ Feature: User should be able to add a comment
   I would like to be able to leave a comment
 
   Background:
-    Given the following articles exist
-    | title               | body                  |
-    | Holger is the best  | Dreamteam for the win |
+    Given following user exist
```
Comment (jyuw) [inline features/user_can_make_comments.feature:7]: And here :P

File: features/user_can_see_articles.feature [hunk 20]
```diff
@@ -4,11 +4,22 @@ Feature: User should be able to see articles
   I would like the articles to be displayed
 
   Background:
-    Given the following articles exist
-    | title               | body                  |
-    | Holger is the best  | Dreamteam for the win |
+    Given following user exist
```
Comment (jyuw) [inline features/user_can_see_articles.feature:7]: lawl

File: features/user_should_be_able_to_create_articles.feature [hunk 21]
```diff
@@ -7,9 +7,13 @@ Feature: User should be able to create articles
     Given the following categories exist
       | name    |
       | Fashion |
+    And following user exist
```
Comment (jyuw) [inline features/user_should_be_able_to_create_articles.feature:10]: I'm the worst soz

File: features/user_should_be_able_to_create_articles.feature [hunk 22]
```diff
@@ -7,9 +7,13 @@ Feature: User should be able to create articles
...
+      | email                 | role       |
+      | author@email.com      | author     |
+    And I am logged in as 'author@email.com'
 
   Scenario: User successfully creates an article
-    Given I am on the "Create Article" page
+    And I am on the "Create Article" page
```
Comment (jyuw) [inline features/user_should_be_able_to_create_articles.feature:19]: Should just be `Given` I'm pretty sure

File: features/user_should_be_able_to_edit_articles.feature [hunk 23]
```diff
@@ -4,18 +4,20 @@ Feature: User should have the ability to edit articles
   I would like to be able to edit an article
 
   Background:
-    Given the following articles exist
+    Given following user exist
```
Comment (jyuw) [inline features/user_should_be_able_to_edit_articles.feature:7]: Last one!

File: spec/policies/article_policy_spec.rb [hunk 24]
```diff
@@ -0,0 +1,34 @@
+require 'rails_helper'
+
+RSpec.describe ArticlePolicy do
+  subject { described_class.new(user, article) }
+
+  let(:article) { create(:article) }
+
+  context 'user is a visitor' do
+    let(:user) { create(:user, role: 'visitor') }
+
```
Comment (jyuw) [inline spec/policies/article_policy_spec.rb]: Extra line break here

File: spec/policies/article_policy_spec.rb [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+    it { is_expected.to permit_new_and_create_actions }
+  end
+
+  context 'user is a subscriber' do
+    let(:user) { create(:user, role: 'subscriber') }
+
+    it { is_expected.to permit_actions [:show] }
+  end
+
```
Comment (jyuw) [inline spec/policies/article_policy_spec.rb]: And here

**Review by jyuw — COMMENTED** [review]
Looks good guys (: just minor formatting stuff

File: app/policies/application_policy.rb [hunk 26]
```diff
@@ -0,0 +1,53 @@
...
+    @user = user
+    @record = record
+  end
+
+  def index?
+    false
+  end
+
+  def show?
+    scope.where(:id => record.id).exists?
```
Comment (magnus-thor) [inline app/policies/application_policy.rb]: Old syntax. use `scope.where(id: record.id).exists?`

File: app/models/user.rb [hunk 27]
```diff
@@ -1,4 +1,20 @@
 class User < ApplicationRecord
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :trackable, :validatable
+
```
Comment (magnus-thor) [inline app/models/user.rb]: In here you can setup a validation of the roles.
`validates_inclusion_of :roles, in: [ 'admin',  'author', 'visitor', 'subscriber']`
then in your unit test you can do
`it { is_expected.to validate_inclusion_of(:role).in_array(['admin',  'author', 'visitor', 'subscriber']) }`

File: features/step_definitions/basic_steps.rb [hunk 28]
```diff
@@ -31,11 +36,22 @@
...
+  visit find_article(article_title)
+end
+
+Given("I try to visit the {string} page") do |string|
+  visit new_article_path
```
Comment (magnus-thor) [inline features/step_definitions/basic_steps.rb]: you have an unused variable `string`

**Review by jyuw — DISMISSED** [review]
Looks good guys (:

**Review by magnus-thor — COMMENTED** [review]
It looks good, But get @tochman to do a review

**Review by jyuw — DISMISSED** [review]

File: Gemfile [hunk 29]
```diff
@@ -3,34 +3,38 @@ git_source(:github) { |repo| "https://github.com/#{repo}.git" }
...
+  gem 'pundit-matchers', '~> 1.4.1'
   gem 'rspec-rails'
+  gem 'selenium-webdriver'
   gem 'shoulda-matchers'
-  gem 'cucumber-rails', require: false
-  gem 'database_cleaner'
-  gem 'launchy'
-  gem 'coveralls', require: false
+  gem 'chromedriver-helper'
+  gem 'selenium-webdriver'
```
Comment (tochman) [inline Gemfile]: This is a duplicate

File: Gemfile [hunk 30]
```diff
@@ -3,34 +3,38 @@ git_source(:github) { |repo| "https://github.com/#{repo}.git" }
...
 end
 
 group :development do
-  gem 'web-console', '>= 3.3.0'
   gem 'listen', '>= 3.0.5', '< 3.2'
   gem 'spring'
   gem 'spring-watcher-listen', '~> 2.0.0'
+  gem 'web-console', '>= 3.3.0'
```
Comment (tochman) [inline Gemfile:37]: Why do you add this dependency?

File: app/controllers/articles_controller.rb [hunk 31]
```diff
@@ -1,9 +1,21 @@
 class ArticlesController < ApplicationController
   before_action :find_article_and_category, except: [:create]
 
+  def new
+    @article = Article.new
+    authorize @article
+  end
+
+  def show
+    @article = Article.find_by(id: params[:id])
```
Comment (tochman) [inline app/controllers/articles_controller.rb]: If you have the `before_action :find_article_and_category` you don't need to assign `@article` again, right?

File: app/controllers/articles_controller.rb [hunk 32]
```diff
@@ -37,12 +51,12 @@ def update
...
-      categories.each do |category_id|
-        category = Category.find_by(id: category_id)
-        @article.categories.include?(category) ? next : @article.categories << category
-      end
+    categories = params[:article][:categories]
+    categories.each do |category_id|
+      category = Category.find_by(id: category_id)
+      @article.categories.include?(category) ? next : @article.categories << category
     end
+  end
```
Comment (tochman) [inline app/controllers/articles_controller.rb:59]: This is why thoroughly performed code reviews are important. By merging badly formatted code in a previous PR and fixing it now, you are clouting this PR and making it harder to review. Please set higher standards for the team.

File: app/models/user.rb [hunk 33]
```diff
@@ -1,4 +1,21 @@
...
+    role == 'visitor'
+  end
+
+  def admin?
+    role == 'admin'
+  end
+
+  def subscriber?
+    role == 'subscriber'
+  end
```
Comment (tochman) [inline app/models/user.rb]: There's another way of dealing with roles that I showcased in my guide (using `enum`) that would NOT require creating a separate method for each role.

File: db/migrate/20180320160301_add_role_to_users.rb [hunk 34]
```diff
@@ -0,0 +1,5 @@
+class AddRoleToUsers < ActiveRecord::Migration[5.2]
+  def change
+    add_column :users, :role, :string, default: 'visitor'
```
Comment (tochman) [inline db/migrate/20180320160301_add_role_to_users.rb]: Please look into the `enum` method in my Role-based guide.

File: spec/policies/article_policy_spec.rb [hunk 35]
```diff
@@ -0,0 +1,31 @@
...
+  context 'user is a visitor' do
+    let(:user) { create(:user, role: 'visitor') }
+
+    it { is_expected.to forbid_new_and_create_actions }
+  end
+
+  context 'user is an author' do
+    let(:user) { create(:user, role: 'author') }
+
```
Comment (tochman) [inline spec/policies/article_policy_spec.rb]: No need for the extra line breaks.

File: spec/spec_helper.rb [hunk 36]
```diff
@@ -1,3 +1,4 @@
+require 'pundit/matchers'
```
Comment (tochman) [inline spec/spec_helper.rb]: Please make sure that Coveralls stays on top. That is a requirement.

**Review by tochman — CHANGES_REQUESTED** [review]

File: Gemfile (hunk 30 above)
Comment (davidboll) [reply Gemfile:37]: We didn't add it, we just sorted the gem's in alphabetical order 👍

File: app/controllers/articles_controller.rb (hunk 31 above)
Comment (davidboll) [reply app/controllers/articles_controller.rb]: Ah good point!

File: app/models/user.rb (hunk 33 above)
Comment (davidboll) [reply app/models/user.rb]: We will look into that, thank you!

File: spec/policies/article_policy_spec.rb (hunk 35 above)
Comment (davidboll) [reply spec/policies/article_policy_spec.rb]: We'll take care of this straight away 👍

File: spec/spec_helper.rb (hunk 36 above)
Comment (davidboll) [reply spec/spec_helper.rb]: Will do 👍

File: app/controllers/articles_controller.rb (hunk 32 above)
Comment (davidboll) [reply app/controllers/articles_controller.rb:59]: We will do better, thanks for pointing this out!

File: Gemfile (hunk 30 above)
Comment (tochman) [reply Gemfile:37]: Okay

File: app/controllers/application_controller.rb [hunk 37]
```diff
@@ -1,2 +1,10 @@
 class ApplicationController < ActionController::Base
+  include Pundit
+
+  rescue_from Pundit::NotAuthorizedError, with: :user_not_authorized
+
+  def user_not_authorized
+    redirect_to root_path, notice: 'Aja baja!'
+  end
 end
+# user = User.new({password: token, email: "#{token}@example.com"}.merge!(@coordinates))
```
Comment (tochman) [inline app/controllers/application_controller.rb]: Please remove comments

File: app/controllers/articles_controller.rb (hunk 32 above)
Comment (tochman) [reply app/controllers/articles_controller.rb:59]: Okay, thx.

File: app/controllers/application_controller.rb (hunk 37 above)
Comment (davidboll) [reply app/controllers/application_controller.rb]: Will do!

**Review by larathordar — DISMISSED** [review]
Looks great !

**Review by larathordar — APPROVED** [review]
good job

**Review by tochman — APPROVED** [review]

**Pull request merged**